### PR TITLE
Refactor template deployment cmdlets.

### PR DESCRIPTION
**Pull request by Tiano2017**
## Checklist

- [ ] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md)
- [ ] The title of the PR is clear and informative
- [ ] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#updating-the-change-log)
- [ ] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (Microsoft internal partners only)
    - [ ] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/help-generation.md)

### CODE REVIEW

**Comment by Tiano2017** [comment]
@cormacpayne a few test cases are failing. I tried to run them in master, and failing as well. Anything I need to do about them?

File: src/Resources/ResourceManager/Components/Constants.cs [hunk 1]
```diff
@@ -30,9 +30,14 @@ public static class Constants
...
-        /// The <c>Microsoft.Management</c> namespace.
+        /// The management group type.
         /// </summary>
-        public static readonly string MicrosoftManagementNamespace = "Microsoft.Management";
+        public static readonly string ManagementGroupType = "/providers/Microsoft.Management/managementGroups/";
+
+        /// <summary>
+        /// The management group type name.
+        /// </summary>
+        public static readonly string ManagementGroupTypeName = "managementGroups";
```
Comment (majastrz) [inline src/Resources/ResourceManager/Components/Constants.cs]: >ManagementGroupTypeName [](start = 38, length = 23)

nit - from the existing naming convention, this should be called `ManagementGroups`

File: src/Resources/ResourceManager/Components/Constants.cs [hunk 2]
```diff
@@ -147,7 +152,7 @@ public static class Constants
         /// <summary>
         /// The management group resource type.
         /// </summary>
-        public static readonly string MicrosoftManagementGroupDefinitionType = Constants.MicrosoftManagementNamespace + "/managementgroups";
+        public static readonly string MicrosoftManagementGroupDefinitionType = Constants.ManagementGroupTypeName + "/managementgroups";
```
Comment (majastrz) [inline src/Resources/ResourceManager/Components/Constants.cs]: >Definition [](start = 62, length = 10)

nit - I think you should drop "definition" from the name. There's really no concept of "Definition" in MGs

File: src/Resources/ResourceManager/Properties/Resources.resx [hunk 3]
```diff
@@ -144,6 +144,9 @@
   <data name="DeploymentDoesntExistAtSubscriptionScope" xml:space="preserve">
     <value>Deployment '{0}' doesn't exist in subscription.</value>
   </data>
+  <data name="DeploymentDoesntExistAtManagementGroupScope" xml:space="preserve">
+    <value>Deployment '{0}' doesn't exist in management group.</value>
```
Comment (majastrz) [inline src/Resources/ResourceManager/Properties/Resources.resx]: >group [](start = 56, length = 5)

Should we mention which MG it is?

File: src/Resources/ResourceManager/Properties/Resources.resx [hunk 4]
```diff
@@ -144,6 +144,9 @@
   <data name="DeploymentDoesntExistAtSubscriptionScope" xml:space="preserve">
     <value>Deployment '{0}' doesn't exist in subscription.</value>
```
Comment (majastrz) [inline src/Resources/ResourceManager/Properties/Resources.resx]: >subscription [](start = 45, length = 12)

You didn't change this, but should we include the subscription ID in the error?

File: src/Resources/ResourceManager/Properties/Resources.resx [hunk 5]
```diff
@@ -396,8 +402,17 @@
   <data name="NoRunningDeployments" xml:space="preserve">
     <value>There are no running deployments under resource group '{0}'</value>
   </data>
+  <data name="NoRunningDeploymentsAtTenantScope" xml:space="preserve">
+    <value>There is no running deployment called '{0}' at the current tenant scope.</value>
```
Comment (majastrz) [inline src/Resources/ResourceManager/Properties/Resources.resx]: >scope [](start = 77, length = 5)

Should we include the tenant ID?

File: src/Resources/Resources.Test/Resources.Test.csproj [hunk 6]
```diff
@@ -24,10 +24,25 @@
     <ProjectReference Include="..\Resources\Resources.csproj" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Reference Include="System">
+      <HintPath>System</HintPath>
```
Comment (majastrz) [inline src/Resources/Resources.Test/Resources.Test.csproj]: > <HintPath>System</HintPath> [](start = 5, length = 28)

Are these valid hint paths?

File: src/Resources/ResourceManager/Components/Constants.cs (hunk 1 above)
Comment (Tiano2017) [reply src/Resources/ResourceManager/Components/Constants.cs]: refactored.

File: src/Resources/ResourceManager/Components/Constants.cs (hunk 2 above)
Comment (Tiano2017) [reply src/Resources/ResourceManager/Components/Constants.cs]: did some refactoring and removed this constant.

File: src/Resources/ResourceManager/Properties/Resources.resx (hunk 3 above)
Comment (Tiano2017) [reply src/Resources/ResourceManager/Properties/Resources.resx]: we should. updated.

File: src/Resources/ResourceManager/Properties/Resources.resx (hunk 4 above)
Comment (Tiano2017) [reply src/Resources/ResourceManager/Properties/Resources.resx]: For all the subscription level scenarios, we work on the current subscription in the context. same for tenant. We could put the subscription Id in the message, but my feeling is it's not necessary. Also, I'm not sure how to get the current tenant Id. :)

File: src/Resources/ResourceManager/Properties/Resources.resx (hunk 5 above)
Comment (Tiano2017) [reply src/Resources/ResourceManager/Properties/Resources.resx]: replied in another comment.

File: src/Resources/Resources.Test/Resources.Test.csproj (hunk 6 above)
Comment (Tiano2017) [reply src/Resources/Resources.Test/Resources.Test.csproj]: not sure how these changes got here. reverted.

File: src/Resources/ResourceManager/Properties/Resources.resx (hunk 4 above)
Comment (majastrz) [reply src/Resources/ResourceManager/Properties/Resources.resx]: Ok, sounds good.

---
In reply to: [307948665](https://github.com/Azure/azure-powershell/pull/9707#discussion_r307948665) [](ancestors = 307948665)

File: src/Resources/ResourceManager/Properties/Resources.resx (hunk 5 above)
Comment (majastrz) [reply src/Resources/ResourceManager/Properties/Resources.resx]: Yup, makes sense. They can do Get-AzContext or something like that.

---
In reply to: [307948673](https://github.com/Azure/azure-powershell/pull/9707#discussion_r307948673) [](ancestors = 307948673)

File: src/Resources/ResourceManager/Implementation/CmdletBase/DeploymentCmdletWithParameters.cs [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+            Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = "Uri or local path to the template file.")]
+        [Parameter(ParameterSetName = TenantParameterSetWithTemplateFileParameterObject,
+            Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = "Uri or local path to the template file.")]
+        [Parameter(ParameterSetName = TenantParameterSetWithTemplateFileParameterFile,
+            Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = "Uri or local path to the template file.")]
+        [Parameter(ParameterSetName = TenantParameterSetWithParameterlessTemplateFile,
+            Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = "Uri or local path to the template file.")]
+        [ValidateNotNullOrEmpty]
+        [Alias("TemplateUri")]
+        public override string TemplateFile { get; set; }
```
Comment (majastrz) [inline src/Resources/ResourceManager/Implementation/CmdletBase/DeploymentCmdletWithParameters.cs]: >TemplateFile [](start = 31, length = 12)

I checked TemplateFile appears in the name of each parameter set name constant and it seems to be (same for the other permuations), but this stuff is pretty much write-only.

Is the existing test coverage good enough to confirm whether we're going to cause any back compat issues? (I don't see an issue, but who knows?)

File: src/Resources/ResourceManager/Implementation/Deployments/RemoveAzureDeploymentCmdlet.cs [hunk 8]
```diff
@@ -69,17 +86,91 @@ public override void ExecuteCmdlet()
...
+                ResourceManagerSdkClient.DeleteDeploymentAtManagementGroup(managementGroupId, deploymentName);
+            }
+            else if (!string.IsNullOrEmpty(resourceGroupName))
+            {
+                ResourceManagerSdkClient.DeleteDeploymentAtResourceGroup(resourceGroupName, deploymentName);
+            }
+            else
+            {
+                ResourceManagerSdkClient.DeleteDeploymentAtSubscriptionScope(deploymentName);
+            }
```
Comment (majastrz) [inline src/Resources/ResourceManager/Implementation/Deployments/RemoveAzureDeploymentCmdlet.cs]: Not necessarily an issue, but I'm seeing this switch pattern in a bunch of places. Would it make sense to capture it in a method that takes a tenant-level delegate, a mg-level delegate, a sub-level delegate, and an RG-level delegate? With named parameters, it might help readability. 

On the other hand, once you figure it out one of them, it's obvious what this does. What do you think?

File: src/Resources/ResourceManager/Implementation/Deployments/SaveAzureDeploymentTemplateCmdlet.cs [hunk 9]
```diff
@@ -87,5 +109,65 @@ protected override void OnProcessRecord()
...
+            {
+                return this.ResourceGroupName;
+            }
+            else if (this.DeploymentObject != null)
+            {
+                return this.DeploymentObject.ResourceGroupName;
+            }
+
+            return null;
+        }
```
Comment (majastrz) [inline src/Resources/ResourceManager/Implementation/Deployments/SaveAzureDeploymentTemplateCmdlet.cs]: Are some of these repeated? Should we throw them into some base class?

File: src/Resources/ResourceManager/Implementation/Deployments/TestAzureDeploymentCmdlet.cs [hunk 10]
```diff
@@ -12,38 +12,152 @@
...
         {
+            if (string.IsNullOrEmpty(this.ResourceGroupName) && this.Mode == DeploymentMode.Complete)
+            {
+                WriteExceptionError(new ArgumentException(ProjectResources.InvalidDeploymentMode));
+            }
+
+            if (this.RollbackToLastDeployment && !string.IsNullOrEmpty(this.RollBackDeploymentName))
+            {
+                WriteExceptionError(new ArgumentException(ProjectResources.InvalidRollbackParameters));
+            }
```
Comment (majastrz) [inline src/Resources/ResourceManager/Implementation/Deployments/TestAzureDeploymentCmdlet.cs]: The backend will respond with a 400 in these situation right? Do we even need these?

Comment (majastrz) [reply src/Resources/ResourceManager/Implementation/Deployments/TestAzureDeploymentCmdlet.cs]: Back-compat maybe?

---
In reply to: [308505704](https://github.com/Azure/azure-powershell/pull/9707#discussion_r308505704) [](ancestors = 308505704)

**Comment by cormacpayne** [comment]
@Tiano2017 would you mind pulling the latest from the `master` branch and resolving the merge conflicts?

File: src/Resources/ResourceManager/Implementation/CmdletBase/DeploymentCmdletWithParameters.cs (hunk 7 above)
Comment (Tiano2017) [reply src/Resources/ResourceManager/Implementation/CmdletBase/DeploymentCmdletWithParameters.cs]: I think we have good coverage on New-AzResourceGroupDeployment, but not for New-AzDeployment. I'll see if I can add some tests there.

File: src/Resources/ResourceManager/Implementation/Deployments/RemoveAzureDeploymentCmdlet.cs (hunk 8 above)
Comment (Tiano2017) [reply src/Resources/ResourceManager/Implementation/Deployments/RemoveAzureDeploymentCmdlet.cs]: I'm usually not a big fan of passing delegates to functions unless it really simplifies implementation or improves readability.  I can take a look and see if that's the case here.

File: src/Resources/ResourceManager/Implementation/Deployments/SaveAzureDeploymentTemplateCmdlet.cs (hunk 9 above)
Comment (Tiano2017) [reply src/Resources/ResourceManager/Implementation/Deployments/SaveAzureDeploymentTemplateCmdlet.cs]: sure. will take a look.

File: src/Resources/ResourceManager/Implementation/Deployments/TestAzureDeploymentCmdlet.cs (hunk 10 above)
Comment (Tiano2017) [reply src/Resources/ResourceManager/Implementation/Deployments/TestAzureDeploymentCmdlet.cs]: Right. it's for back-compat, making sure for resource group level it works the same as New-AzResourceGroupDeployment.

File: src/Resources/ResourceManager/Implementation/Deployments/RemoveAzureDeploymentCmdlet.cs (hunk 8 above)
Comment (majastrz) [reply src/Resources/ResourceManager/Implementation/Deployments/RemoveAzureDeploymentCmdlet.cs]: Sounds good. I'm ok either way.

---
In reply to: [308965056](https://github.com/Azure/azure-powershell/pull/9707#discussion_r308965056) [](ancestors = 308965056)

File: src/Resources/ResourceManager/Implementation/Deployments/TestAzureDeploymentCmdlet.cs (hunk 10 above)
Comment (majastrz) [reply src/Resources/ResourceManager/Implementation/Deployments/TestAzureDeploymentCmdlet.cs]: Makes sense.

---
In reply to: [308965884](https://github.com/Azure/azure-powershell/pull/9707#discussion_r308965884) [](ancestors = 308965884)

File: src/Resources/ResourceManager/SdkClient/ResourceManagerSdkClient.cs [hunk 11]
```diff
@@ -851,111 +1074,384 @@ public virtual List<PSResourceGroupDeployment> FilterResourceGroupDeployments(Fi
...
+        /// List deployment operations at tenant scope.
+        /// </summary>
+        /// <param name="deploymentName">The deployment name</param>
+        /// <param name="operationId">The operation Id</param>
+        /// <returns>The deployment operations</returns>
+        public virtual List<PSDeploymentOperation> ListDeploymentOperationsAtTenantScope(string deploymentName, string operationId = null)
+        {
+            List<PSDeploymentOperation> deploymentOperations = new List<PSDeploymentOperation>();
+
+            // (tiano): to do.
```
Comment (majastrz) [inline src/Resources/ResourceManager/SdkClient/ResourceManagerSdkClient.cs]: >// (tiano): to do. [](start = 12, length = 18)

This will be a separate PR?

**Comment by Tiano2017** [comment]
@cormacpayne can you help take another look at this PR?

File: src/Resources/Resources/help/Get-AzDeployment.md [hunk 12]
```diff
@@ -12,41 +12,89 @@ Get deployment
...
+PS C:\>Get-AzDeployment -ScopeType Subscription -Name "DeployRoles01"
 ```
 
 This command gets the DeployRoles01 deployment at the current subscription scope.
 You can assign a name to a deployment when you create it by using the **New-AzDeployment** cmdlets.
 If you do not assign a name, the cmdlets provide a default name based on the template that is used to create the deployment.
 
+### Example 3: Get a deployment by deployment ID
+```
+PS C:\>Get-AzDeployment -Id "/subscriptions/sub-01/providers/Microsoft.Resources/deployments/DeployRole01"
```
Comment (cormacpayne) [inline src/Resources/Resources/help/Get-AzDeployment.md]: @Tiano2017 since we're adding examples, can we add the corresponding output from running each command in all six of the examples in this file? This will give users an idea of what is returned when the cmdlet is run

File: src/Resources/Resources/help/Get-AzDeploymentOperation.md [hunk 13]
```diff
@@ -35,16 +55,37 @@ For more on this setting and how to enable it, see New-AzDeployment and Debuggin
...
 ```
-PS C:\>Get-AzDeploymentOperation -DeploymentName test
+PS C:\>Get-AzDeploymentOperation -ScopeType "Subscription" -DeploymentName test
 ```
 
 Gets deployment operation with name "test" at the current subscription scope.
 
+### Get deployment operations for a deployment at a resource group given a deployment name
+```
+PS C:\>Get-AzDeploymentOperation -ScopeType "ResourceGroup" -DeploymentName test -ResourceGroupName "testrg"
```
Comment (cormacpayne) [inline src/Resources/Resources/help/Get-AzDeploymentOperation.md]: @Tiano2017 same comment about adding output to all of the examples in this file

File: src/Resources/Resources/help/New-AzDeployment.md [hunk 14]
```diff
@@ -124,26 +198,52 @@ Template parameter values that you enter at the command prompt take precedence o
 
 ## EXAMPLES
 
-### Example 1: Use a custom template and parameter file to create a deployment
+### Example 1: Create a deployment at subscription scope with a custom template and parameter file
 ```
-PS C:\> New-AzDeployment -Location "West US" -TemplateFile "D:\Azure\Templates\EngineeringSite.json" -TemplateParameterFile "D:\Azure\Templates\EngSiteParms.json" -TemplateVersion "2.1"
+PS C:\> New-AzDeployment -ScopeType "Subscription" -DeploymentName "deploy-01" -Location "West US" -TemplateFile "D:\Azure\Templates\ServiceTemplate.json" -TemplateParameterFile "D:\Azure\Templates\ServiceParameters.json"
```
Comment (cormacpayne) [inline src/Resources/Resources/help/New-AzDeployment.md]: @Tiano2017 same comment about adding the corresponding output to all of the examples in this file

File: src/Resources/Resources/help/Remove-AzDeployment.md [hunk 15]
```diff
@@ -31,20 +53,41 @@ Remove-AzDeployment -InputObject <PSDeployment> [-AsJob] [-PassThru] [-ApiVersio
...
-The **Remove-AzDeployment** cmdlet removes an Azure deployment at subscription scope and any associated operations.
+The **Remove-AzDeployment** cmdlet removes an Azure deployment and its associated operations at subscription, resource group, management group, or tenant scope.
 
 ## EXAMPLES
 
-### Example 1: Remove a deployment with a given name
+### Example 1: Remove a deployment at subscription scope
 ```
-PS C:\>Remove-AzDeployment -Name "RolesDeployment"
+PS C:\>Remove-AzDeployment -ScopeType "Subscription" -Name "RolesDeployment"
```
Comment (cormacpayne) [inline src/Resources/Resources/help/Remove-AzDeployment.md]: @Tiano2017 same comment about adding the corresponding output to all examples in this file (if this cmdlet does return any output)

File: src/Resources/Resources/help/Save-AzDeploymentTemplate.md [hunk 16]
```diff
@@ -25,20 +47,20 @@ Save-AzDeploymentTemplate -DeploymentObject <PSDeployment> [-Path <String>] [-Fo
...
-The **Save-AzDeploymentTemplate**  cmdlet saves a deployment template to a JSON file.
+The **Save-AzDeploymentTemplate** cmdlet saves a deployment template to a JSON file.
 
 ## EXAMPLES
 
-### Example 1
+### Example 1 Save template for a deployment at subscription scope
 ```powershell
-PS C:\> Save-AzDeploymentTemplate -DeploymentName "TestDeployment"
+PS C:\> Save-AzDeploymentTemplate -ScopeType "Subscription" -DeploymentName "TestDeployment"
```
Comment (cormacpayne) [inline src/Resources/Resources/help/Save-AzDeploymentTemplate.md]: @Tiano2017 same comment about adding the corresponding output to all examples in this file (if this cmdlet does return anything)

File: src/Resources/Resources/help/Stop-AzDeployment.md [hunk 17]
```diff
@@ -42,14 +62,35 @@ This cmdlet stops only one running deployment. Use the *Name* parameter to stop
 
 ### Example 1
 ```
-PS C:\>Stop-AzDeployment -Name "deployment01"
+PS C:\>Stop-AzDeployment -ScopeType "Subscription" -Name "deployment01"
```
Comment (cormacpayne) [inline src/Resources/Resources/help/Stop-AzDeployment.md]: @Tiano2017 same comment about adding the corresponding output to all examples in this file (if this cmdlet does return anything)

File: src/Resources/Resources/help/Test-AzDeployment.md [hunk 18]
```diff
@@ -12,104 +12,188 @@ Validates a deployment.
...
 ## DESCRIPTION
 The **Test-AzDeployment** cmdlet determines whether a deployment template and its parameter values are valid.
 
 ## EXAMPLES
 
-### Example 1: Test deployment with a custom template and parameter file
+### Example 1: Test deployment at subscription scope with a custom template and parameter file
 ```
-PS C:\> Test-AzDeployment -Location "West US" -TemplateFile "D:\Azure\Templates\EngineeringSite.json" -TemplateParameterFile "D:\Azure\Templates\EngSiteParms.json"
+PS C:\> Test-AzDeployment -ScopeType "Subscription" -Location "West US" -TemplateFile "D:\Azure\Templates\EngineeringSite.json" -TemplateParameterFile "D:\Azure\Templates\EngSiteParms.json"
```
Comment (cormacpayne) [inline src/Resources/Resources/help/Test-AzDeployment.md]: @Tiano2017 same comment about adding the corresponding output to all examples in this file

File: tools/StaticAnalysis/Exceptions/Az.Resources/BreakingChangeIssues.csv [hunk 19]
```diff
@@ -109,4 +109,12 @@
 "Microsoft.Azure.PowerShell.Cmdlets.ResourceManager.Cmdlets.dll","Microsoft.Azure.Commands.ResourceManager.Cmdlets.Implementation.SetAzurePolicyAssignmentCmdlet","Set-AzPolicyAssignment","0","1050","The parameter set 'NameParameterSet' for cmdlet 'Set-AzPolicyAssignment' has been removed.","Add parameter set 'NameParameterSet' back to cmdlet 'Set-AzPolicyAssignment'."
```
Comment (cormacpayne) [inline tools/StaticAnalysis/Exceptions/Az.Resources/BreakingChangeIssues.csv]: @Tiano2017 when is it planned to publish version 2.3.0-preview of `Micorosft.Azure.Management.ResourceManager` to NuGet?

File: src/Resources/Resources.Test/ScenarioTests/DeploymentScopeTests.cs [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+        }
+
+        [Fact()]
+        [Trait(Category.AcceptanceType, Category.CheckIn)]
+        public void SubscriptionLevelDeploymentEndToEnd()
+        {
+            ResourcesController.NewInstance.RunPsTest(_logger, "Test-DeploymentEndToEnd-SubscriptionScope");
+        }
+
+        [Fact(Skip = "Job tests are flaky.")]
```
[the remark this answers is not in the document]
Comment (Tiano2017) [reply src/Resources/Resources.Test/ScenarioTests/DeploymentScopeTests.cs]: I couldn't record it as the test case run cmdlets with "-AsJob". I'll just remove them.

File: src/Resources/Resources.Test/ScenarioTests/DeploymentScopeTests.ps1 [hunk 21]
```diff
@@ -0,0 +1,245 @@
...
+		Assert-NotNull $templatePath.Path
+
+		$operations = Get-AzDeploymentOperation -DeploymentName $deploymentName
+		Assert-AreEqual 5 @($operations).Count
+
+		Remove-AzDeployment -Name $deploymentName
+	}
+	finally
+	{
+
```
[the remark this answers is not in the document]
Comment (Tiano2017) [reply src/Resources/Resources.Test/ScenarioTests/DeploymentScopeTests.ps1]: not a big issue but I think we can remove it. Will update.

File: src/Resources/ResourceManager/Implementation/Deployments/GetAzureDeploymentCmdlet.cs [hunk 22]
```diff
@@ -12,49 +12,144 @@
...
+        internal const string ManagementGroupParameterSetWithDeploymentName = "ManagementGroupWithDeploymentName";
+        internal const string TenantParameterSetWithDeploymentName = "TenantWithDeploymentName";
+
         internal const string DeploymentIdParameterSet = "GetByDeploymentId";
 
-        /// <summary>
-        /// The deployment name parameter set.
-        /// </summary>
-        internal const string DeploymentNameParameterSet = "GetByDeploymentName";
+        [Parameter(Position = 0, ParameterSetName = GetAzureDeploymentCmdlet.ResourceGroupParameterSetWithDeploymentName, Mandatory = true,
```
[the remark this answers is not in the document]
Comment (Tiano2017) [reply src/Resources/ResourceManager/Implementation/Deployments/GetAzureDeploymentCmdlet.cs]: Right. we did some research and decided to require `-ScopeType` for all scopes.

File: tools/StaticAnalysis/Exceptions/Az.Resources/BreakingChangeIssues.csv (hunk 19 above)
Comment (Tiano2017) [reply tools/StaticAnalysis/Exceptions/Az.Resources/BreakingChangeIssues.csv]: we'll publish it in the next few days. Actually the new release is going to be 2.4.0-preview.

File: src/Resources/Resources.Test/ScenarioTests/DeploymentScopeTests.cs [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+        }
+
+        [Fact()]
+        [Trait(Category.AcceptanceType, Category.CheckIn)]
+        public void ResourceGroupDeploymentEndToEnd()
+        {
+            ResourcesController.NewInstance.RunPsTest(_logger, "Test-DeploymentEndToEnd-ResourceGroup");
+        }
+
+        [Fact(Skip = "Job tests are flaky.")]
```
[the remark this answers is not in the document]
Comment (Tiano2017) [reply src/Resources/Resources.Test/ScenarioTests/DeploymentScopeTests.cs]: addressed.

File: src/Resources/Resources.Test/ScenarioTests/DeploymentScopeTests.cs [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+        }
+
+        [Fact()]
+        [Trait(Category.AcceptanceType, Category.CheckIn)]
+        public void ManagementGroupLevelDeploymentEndToEnd()
+        {
+            ResourcesController.NewInstance.RunPsTest(_logger, "Test-DeploymentEndToEnd-ManagementGroup");
+        }
+
+        [Fact(Skip = "Job tests are flaky.")]
```
[the remark this answers is not in the document]
Comment (Tiano2017) [reply src/Resources/Resources.Test/ScenarioTests/DeploymentScopeTests.cs]: addressed.

File: src/Resources/Resources.Test/ScenarioTests/DeploymentScopeTests.ps1 [hunk 25]
```diff
@@ -0,0 +1,245 @@
...
+		Assert-AreEqual $getById.DeploymentName $deployment.DeploymentName
+
+		$operations = Get-AzDeploymentOperation -DeploymentName $deploymentName
+		Assert-AreEqual 5 @($operations).Count
+
+		Remove-AzDeployment -Name $deploymentName
+	}
+	finally
+	{
+
```
[the remark this answers is not in the document]
Comment (Tiano2017) [reply src/Resources/Resources.Test/ScenarioTests/DeploymentScopeTests.ps1]: updated.

File: src/Resources/Resources.Test/ScenarioTests/DeploymentScopeTests.ps1 [hunk 26]
```diff
@@ -0,0 +1,245 @@
...
+		$templatePath = Save-AzDeploymentTemplate -ManagementGroupId "tiano-mgtest01" -Name $deploymentName -Force
+		Assert-NotNull $templatePath.Path
+		
+		$operations = Get-AzDeploymentOperation -ManagementGroup "tiano-mgtest01" -DeploymentName $deploymentName
+		Assert-AreEqual 4 @($operations).Count
+		#
+		Remove-AzDeployment -ManagementGroup "tiano-mgtest01" -Name $deploymentName
+	}
+	finally
+	{
```
[the remark this answers is not in the document]
Comment (Tiano2017) [reply src/Resources/Resources.Test/ScenarioTests/DeploymentScopeTests.ps1]: updated.

File: src/Resources/Resources.Test/ScenarioTests/DeploymentScopeTests.ps1 [hunk 27]
```diff
@@ -0,0 +1,245 @@
...
+function Test-DeploymentEndToEnd-ManagementGroup
+{
+    try
+	{
+        # Setup
+		$deploymentName = Get-ResourceName
+		$location = "EastUS"
+
+		# Test
+		$deployment = New-AzDeployment -ManagementGroupId "tiano-mgtest01" -Name $deploymentName -Location $location -TemplateFile management_group_level_template.json -TemplateParameterFile management_group_level_parameters.json
```
[the remark this answers is not in the document]
Comment (Tiano2017) [reply src/Resources/Resources.Test/ScenarioTests/DeploymentScopeTests.ps1]: updated.

File: src/Resources/Resources.Test/ScenarioTests/DeploymentScopeTests.ps1 [hunk 28]
```diff
@@ -0,0 +1,245 @@
...
+function Test-DeploymentAsJob-ManagementGroup
+{
+    try
+	{
+	    # Setup
+		$deploymentName = Get-ResourceName
+		$location = "EastUS"
+
+		# Test
+		$job = New-AzDeployment -ManagementGroupId "tiano-mgtest01" -Name $deploymentName -Location $location -TemplateFile management_group_level_template.json -TemplateParameterFile management_group_level_parameters.json -AsJob
```
[the remark this answers is not in the document]
Comment (Tiano2017) [reply src/Resources/Resources.Test/ScenarioTests/DeploymentScopeTests.ps1]: updated.

File: src/Resources/Resources/help/Get-AzDeployment.md (hunk 12 above)
Comment (Tiano2017) [reply src/Resources/Resources/help/Get-AzDeployment.md]: Not sure about this. The output is very large. Does it fit in the help file.

**Comment by Tiano2017** [comment]
Cheers. All checks passed finally. :)

File: src/Resources/ResourceManager/Implementation/Deployments/GetAzureDeploymentCmdlet.cs [hunk 29]
```diff
@@ -12,49 +12,144 @@
...
+            HelpMessage = "The name of deployment.")]
+        [Parameter(Position = 2, ParameterSetName = GetAzureDeploymentCmdlet.ResourceGroupParameterSetWithDeploymentName, Mandatory = false,
             HelpMessage = "The name of deployment.")]
         [ValidateNotNullOrEmpty]
         public string Name { get; set; }
 
         [Alias("DeploymentId", "ResourceId")]
-        [Parameter(ParameterSetName = GetAzureDeploymentCmdlet.DeploymentIdParameterSet, Mandatory = false,
+        [Parameter(ParameterSetName = GetAzureDeploymentCmdlet.DeploymentIdParameterSet, Mandatory = true,
             HelpMessage = "The fully qualified resource Id of the deployment. example: /subscriptions/{subId}/providers/Microsoft.Resources/deployments/{deploymentName}")]
```
[the remark this answers is not in the document]
Comment (Tiano2017) [reply src/Resources/ResourceManager/Implementation/Deployments/GetAzureDeploymentCmdlet.cs]: In that sense it's actually breaking. In the past you can provide neither name nor Id, which gets all the deployments at subscription scope. Now it would not be possible, you will be asked to provide name/scopeType if you run the cmdlet without any parameter.

File: src/Resources/ResourceManager/Implementation/Deployments/GetAzureDeploymentOperationCmdlet.cs [hunk 30]
```diff
@@ -14,42 +14,73 @@
...
+        [Parameter(ParameterSetName = GetAzureDeploymentOperationCmdlet.ResourceGroupParameterSetWithDeploymentName, Mandatory = true,
+            HelpMessage = "The scope type of the deployment.")]
+        [Parameter(ParameterSetName = GetAzureDeploymentOperationCmdlet.SubscriptionParameterSetWithDeploymentName, Mandatory = true,
+            HelpMessage = "The scope type of the deployment.")]
+        [Parameter(ParameterSetName = GetAzureDeploymentOperationCmdlet.ManagementGroupParameterSetWithDeploymentName, Mandatory = true,
+            HelpMessage = "The scope type of the deployment.")]
+        [Parameter(ParameterSetName = GetAzureDeploymentOperationCmdlet.TenantParameterSetWithDeploymentName, Mandatory = true,
+            HelpMessage = "The scope type of the deployment.")]
+        [ValidateNotNullOrEmpty]
+        public DeploymentScopeType? ScopeType { get; set; }
```
[the remark this answers is not in the document]
Comment (Tiano2017) [reply src/Resources/ResourceManager/Implementation/Deployments/GetAzureDeploymentOperationCmdlet.cs]: Switch parameter would only be able to indicate true/false. originally we wanted that as we planed to use to tell whether it's subscription or tenant scope. In our subsequent discussions we decided to do a breaking change and release it in a major release, and we decided to just make it consistent in all four scopes, by using this ScopeType parameter.

File: src/Resources/ResourceManager/Implementation/Deployments/RemoveAzureDeploymentCmdlet.cs [hunk 31]
```diff
@@ -23,26 +25,48 @@ namespace Microsoft.Azure.Commands.ResourceManager.Cmdlets.Implementation
...
+        [Parameter(Position = 0, ParameterSetName = RemoveAzureDeploymentCmdlet.ResourceGroupParameterSetWithDeploymentName, Mandatory = true,
+            HelpMessage = "The scope type of the deployment.")]
+        [Parameter(Position = 0, ParameterSetName = RemoveAzureDeploymentCmdlet.SubscriptionParameterSetWithDeploymentName, Mandatory = true,
+            HelpMessage = "The scope type of the deployment.")]
+        [Parameter(Position = 0, ParameterSetName = RemoveAzureDeploymentCmdlet.ManagementGroupParameterSetWithDeploymentName, Mandatory = true,
+            HelpMessage = "The scope type of the deployment.")]
+        [Parameter(Position = 0, ParameterSetName = RemoveAzureDeploymentCmdlet.TenantParameterSetWithDeploymentName, Mandatory = true,
+            HelpMessage = "The scope type of the deployment.")]
+        [ValidateNotNullOrEmpty]
+        public DeploymentScopeType? ScopeType { get; set; }
```
[the remark this answers is not in the document]
Comment (Tiano2017) [reply src/Resources/ResourceManager/Implementation/Deployments/RemoveAzureDeploymentCmdlet.cs]: The parameter is only allowed in certain parameters. In other parameter sets, we want to it to be unspecified.

**Comment by markcowl** [comment]
@tiano2017 Monday is code complete for this release.  If we want this in for the release, we need to have these comments addressed by then.

**Comment by Tiano2017** [comment]
One test case is failing: Microsoft.Azure.Commands.DataLakeStore.Test.ScenarioTests.AdlsAliasTests.TestAdlsFileSystem. It doesn't seem to be related to my change. Is there anything I need to do about it?

**Comment by Tiano2017** [comment]
@erich-wang, @cormacpayne, @markcowl Is anything else I need to do to get this PR merged?

File: src/Resources/ResourceManager/Implementation/Deployments/GetAzureDeploymentOperationCmdlet.cs (hunk 30 above)
[the remark this answers is not in the document]
Comment (markcowl) [reply src/Resources/ResourceManager/Implementation/Deployments/GetAzureDeploymentOperationCmdlet.cs]: The problem is that this doesn't allow users to fall into the pit of success with the parameters - they can select a scope type and provide parameters for a non-matching scope, resulting in a runtime error.  If you use switch parameters, the user falls into the pit of success - they can only provide parameters for the deployment type they have selected.

Since this is a preview, I won't prevent you from doing this, but you will need to match up with the origiinal design review once this goes GA.

**Comment by kilasuit** [comment]
> 
> ## Description
>
>

PR's really need to have a Description filled out so that users like myself can understand **WHY** changes are being made because it's not always easily apparent from the code being added what the Intent is.
@markcowl can you please make sure that your team ensures these are added in future
